first draft of projecta schema

raidun99 opened the pull request.

sbillinge commented on regolith/schemas.py:
```diff
@@ -29,6 +29,7 @@
                 "submitted", "cancelled"]
 REVIEW_RECOMMENDATION = ["reject", "asis", "smalledits", "diffjournal", "majoredits"]
 SERVICE_TYPE = ["profession", "university", "school", "department"]
+PROJECTA_STATUS = ["proposed", "started", "converged", "finished"]
```
I think these are milestone statuses.  We probably also want "cancelled"

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                            "of a larger project."
+        },
+        "_id": {
+            "description": "Unique projecta identifier",
```
the singular of projecta is projectum.  This will need to be changed nearly everywhere...

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+            "type": "string"
+        },
+        "collaborators": {
+            "description": "list of collaborator's id's",
```
"list of collaborators ids.  These are non-group members. These will be dereferenced from the contacts collection.  Group members will be dereferenced from the people collection."

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+            "description": "outline of the deliverable for this projecta",
+            "type": "dict",
+            "schema": {
+                "audience": {"description": "who is this deliverable for",
```
"the target audience for this deliverable"

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                "audience": {"description": "who is this deliverable for",
+                             "required": False,
+                             "type": "list"},
+                #is this the proper form of the due date, the exemplar is ambiguous
```
yes

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                "success_def": {"description": "definition of a successful deliverable",
+                                "required": False,
+                                "type": "string"},
+                "scope": {"description": "",
```
"a list of items that define the scope of the deliverable.  If this is a software release it might be a list of Use Cases that will be satisfied.  If it is a paper it defines what will, and what won't, be described in the paper."

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                "scope": {"description": "",
+                          "required": False,
+                          "type": "list"},
+                "platform": {"description": "description of how and where the audience will access "
```
nicely done.  Please add at the end "For software releases, this may be the computer operating systems that will be supported, or if it will be a web service, etc."

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                                           "not needed for paper submissions",
+                            "required": False,
+                            "type": "list"},
+                "notes": {"description": "deliverable note",
```
"any notes about the deliverable that we want to keep track of"

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+        },
+        "end_date": {
+            "description": "projecta end date, yyyy-mm-dd. "
+                           "Not necessarily the same as deliverable due date",
```
you can delete this line.  Due dates and end dates are never the same. (unless the thing happens to actually be delivered when it was planned to be.....).   btw, I think this may be a concept missing from Yevgeny's mosy schema, but I let him take the lead on this.....each action on the action graph should have a start and an end datetime that records when it took place, but for planning purposes (when it is recipe and not a dish) it is helpful to have a projected start and end time, or maybe more useful is a due-time and a duration.  I could imagine in the future that if actions had durations and sequence we could find algorithms that would build optimal plans (our beloved gantt charts)

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+            "type": "string"
+        },
+        "group_members": {
+            "description": "group member id's working on this project",
```
please add "These will be dereferenced from the people collection."

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                "objective": {"description": "goal of the meeting",
+                              "required": False,
+                              "type": "string"},
+                "audience": {"description": "rolls of audience members "
```
"list of people attending the meeting.  Normally this list is group_members, collaborators, and pi, or some subset of these.  if people are invited who are not already in these groups their names or id's can be added explicitly to the list"

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                           "required": False,
+                           "type": "string",
+                           "eallowed": PROJECTA_STATUS}
+
```
remove blank line

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+            }
+        },
+        "lead": {
+            "description": "id of projecta lead",
```
"the id of the lead student or person for the projectum.  Person details will be dereferenced from the people collection."

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+            "type": "string"
+        },
+        "log_url": {
+            "description": "link to log of meetings",
```
"link to an online document (e.g., Google doc) that is a log of notes and meeting minutes for the projectum"

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+        },
+        "milestones": {
+            "description": "smaller deliverables done by a certain date "
+                           "a series of milestones ends with the deliverable "
```
insert "projectum" before deliverable

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                                                "e.g. 'pi', 'lead'",
+                                 "required": False,
+                                 "type": "list"},
+                    "status": {"description": "proposed, started, finished, converged",
```
ouch.   use an fstring here and say f"status of the milestone from {MILESTONE_STATI}" sot the documentation stays up to date with the code as the code develops (e.g., if we decide to add a milestone status.....

raidun99 replied:
I didn't understand this comment completely, so I likely didn't implement it how you were envisioning.

sbillinge replied:
> 
> 
> I didn't understand this comment completely, so I likely didn't implement it how you were envisioning.

I tried to give a more complete description in the latest comment

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+                               "required": False,
+                               "type": "string",
+                               "eallowed": PROJECTA_STATUS},
+                    "type": {"description": "what kind of deliverable is it "
```
this should be eallowed and we should discuss what to put in that list.  Look through the projectum.yml db in rg-db-local to see what we have used so far.

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+        "pi_id": {"description": "id of the PI",
+                  "required": False,
+                  "type": "string"},
+        "status": {"description": "proposed, started, finished, converged",
```
see above

sbillinge commented on regolith/schemas.py:
```diff
@@ -3620,6 +3621,172 @@
+            "required": False,
+            "type": "string"
+        },
+        "milestones": {
```
the kickoff and deliverables are also kind of milestones, so please harvest text from their descriptions and synchronize (i.e., if you change it based on my comments there, make sure the changes appear here).

I probably made a mistake to have the kickoff field.  It is just a milestone with the same structure.  But I am not ready to refactor all the code to change that yet.

sbillinge reviewed (commented):
Thanks Ruby, great job!!!!! 

Now, I want to be able to build progress reports from this database so I want to add to each milestone something like
```
progress:
  - text: The samples have been synthesized and placed in the sample cupboard.  They turned out well and are blue as expected.
  - figure: <token that dereferences a figure or image in group local storage db>
  - slides_url: <url to slides describing the development, e.g., Google slides url>
 ```
please could you add a couple of these to the exemplar and to the schema?
Thanks so much!

sbillinge commented:
@raidun99 this is close, could we maybe push to get it finished and merged?  thx.

raidun99 commented:
I left a few comments containing questions in the file. Let me know if this is not the best way to communicate questions to you.

While looking through the projecta db, I saw that a handful of deliverables had a type field. That field is not in the exemplar so I didn't add it to the schema. Should it be in the schema?

Cheers!

sbillinge commented on regolith/schemas.py:
```diff
@@ -1481,7 +1487,7 @@
         },
     ],
-    "projecta": {
+    "projectum": {
```
this is a collection, so it should be projecta (plural)

sbillinge replied:
did you miss this one?

sbillinge commented on regolith/schemas.py:
```diff
@@ -3753,38 +3769,49 @@
+                                             "small, non-deliverable to-dos to reach the milestone",
+                              "required": False,
+                              "type": "list"},
+                    "progress": {"description": "text description of progress and observations, "
```
the schema is not quite right here.  It should be a list of dicts and the dict has to have its own subschema

sbillinge replied:
This needs fixing.  Look elsewhere in the schema for a model

sbillinge commented on regolith/schemas.py:
```diff
@@ -3753,38 +3769,49 @@
-                               "type": "string",
-                               "eallowed": PROJECTA_STATUS},
-                    "type": {"description": "what kind of deliverable is it "
+                               "type": "fstring",
```
is this an allowed type?

sbillinge replied:
string

sbillinge commented on regolith/schemas.py:
```diff
@@ -3753,38 +3769,49 @@
-                    "type": {"description": "what kind of deliverable is it "
+                               "type": "fstring",
+                               "eallowed": PROJECTUM_STATUS},
+                    "type": {"description": "what kind of deliverable the milestone is "
```
I am not sure if we can use fstrings in schemas here but if we can we can do it like this:
```
 "type": {"description": f"what kind of deliverable the milestone is. Allowed values are {", ".join(MILESTONE_TYPE)}"
```
Test it to  make sure it works (i.e., try and run `python run_tests.py tests/test_validate.py` from the main regolith directory.  If it does work, we should use this pattern for all the projecta entries that have ``eallowed``

raidun99 replied:
I tested it. It didn't throw up any warnings, so maybe okay?

sbillinge replied:
cool.  Go for it on all of them....

raidun99 replied:
Completed in the latest commit :)

sbillinge commented on regolith/schemas.py:
```diff
@@ -3753,38 +3769,49 @@
                   "required": False,
                   "type": "string"},
-        "status": {"description": "proposed, started, finished, converged",
+        "status": {"description": "proposed, started, finished, converged, cancelled, paused",
```
see above.

sbillinge reviewed (commented):
Nice, thanks Ruby.  Please see the inline comments.  Also, this needs a rebase.  Do you use Pycharm?  The quickest may be a 5 minute zoom call (before 1pm today).

Nearly there.

raidun99 commented:
Sorry, I had lecture and then another meeting. I do use Pycharm. I think I can just follow the group wiki how to on rebasing: https://github.com/Billingegroup/rg-db-group/wiki/GitHub-Rebase-with-PyCharm and https://github.com/Billingegroup/group_git_workflow

sbillinge commented:
yes, try it yourself!!!! :)   I am in meetings now anyway.

On Wed, Jan 27, 2021 at 2:02 PM raidun99 <notifications@github.com> wrote:

> Sorry, I had lecture and then another meeting. I do use Pycharm. I think I
> can just follow the group wiki how to on rebasing:
> https://github.com/Billingegroup/rg-db-group/wiki/GitHub-Rebase-with-PyCharm
> and https://github.com/Billingegroup/group_git_workflow
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-768503469>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOEBVL3JC2K3XU4YHDS4BPLTANCNFSM4WBLXWVA>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

sbillinge commented on regolith/schemas.py:
```diff
@@ -31,6 +31,13 @@
+# kickoff, deliverable, milestones, and the projectum all have a status field, want different options for each?
+PROJECTUM_STATUS = ["proposed", "started", "converged", "finished", "cancelled", "paused"]
+# this is every milestone type I found in the projecta db
+MILESTONE_TYPE = ["pr", "meeting", "other", "paper", "pull request", "release", "email", "handin",
```
nice work!

pr stands for pull request, so please remove pull request (we will have to update the database to make it validate!)

let's also remove (and we will need a PR on the database to correct these)
code (covered by release or pr)
task (not a deliverable)
file (covered by handin?)
finalized_doc (covered by approval)
software (covered by pr etc.)
merged_pr (covered by pr)
finalized_list (covered by approval...see below)
manuscript (covered by paper)
replace approved_file with approval

sbillinge commented on regolith/schemas.py:
```diff
@@ -1527,7 +1534,11 @@
             'status': 'proposed',
-            'type': 'meeting'
+            'type': 'meeting',
+            'progress': [
```
no open bracket

sbillinge commented on regolith/schemas.py:
```diff
@@ -1527,7 +1534,11 @@
-            'type': 'meeting'
+            'type': 'meeting',
+            'progress': [
+              '- text: The samples have been synthesized and placed in the sample cupboard. They turned out well and are blue as expected.',
```
quote next to the t so the - is not quoted

remove comma

sbillinge commented on regolith/schemas.py:
```diff
@@ -3634,69 +3645,72 @@
         },
     },
-    "projecta": {
+    "projectum": {
```
see above

sbillinge reviewed (commented):
few more comments.  Making progress....

The rebase almost worked.  I saw one surprising commit in there, but I think it is ok.

sbillinge reviewed (commented):
next round of review

raidun99 commented on regolith/schemas.py:
```diff
@@ -3822,7 +3835,8 @@
                   "type": "string"},
-        "status": {"description": "proposed, started, finished, converged, cancelled, paused",
+        "status": {"description": f"status of the projectum. "
+                                  "Allowed values are {', '.join(PROJECTUM_STATUS)}",
```
I could see where the brace was missing. I realized that my f string syntax was wrong, so maybe that was the problem?

sbillinge replied:
sorry, my error.  GitHub didn't show the lines below and I found the missing braces down there.  The problem is elsewhere!

raidun99 commented:
Now test_validate_python is failing every time I test it, but I don't think it has anything to do with the f strings.

sbillinge reviewed (commented):
it may be related to the implied line continuation.  When you use that (i.e., breaking the string onto multiple lines) I think you need to put the f at the beginning of each line.  Not sure if that is the problem but worth a try.  Also, look at the stack dump from the failing test to see where and how it failed.  I can try and look at the CI to see too, but it is begin very slow.

sbillinge commented:
the join has to be joining strings too.  I think some of the allowed values may be integers or sthg. e.g.,
```
", ".join([1,2,3])
```
won't work, 
```
", ".join(["1","2","3"])
```
will.   In this case, rather than changing the allowed values just so we can print them you could use
```
", ".join([str(a) for a in [1,2,3]])
```

raidun99 commented:
I think the test is failing because the at least one of the db files doesn't match its schema. It appears that the validate function is the one calling sys.exit

sbillinge commented:
The test runs the schema against the exemplar, not against any db

On Wed, Jan 27, 2021, 9:22 PM raidun99 <notifications@github.com> wrote:

> I think the test is failing because the at least one of the db files
> doesn't match its schema. It appears that the validate function is the one
> calling sys.exit
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-768750141>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUO4MI3SSTCB6HBJJ2LS4DC7FANCNFSM4WBLXWVA>
> .
>

sbillinge commented on regolith/schemas.py:
```diff
@@ -1526,7 +1530,12 @@
-            'type': 'meeting'
+            'type': 'meeting',
+            'progress':[{
+              - 'text': 'The samples have been synthesized and placed in the sample cupboard. '
```
I was finally able to check the error message from Travis and it is here.  You are mixing json and yaml syntax here.  You need to properly format the dict.  line-breaks aside, what you have is `{- "text": "the samples...." - "figure": "<token..."}`.  A properly composed dict would not have the dashes, and would have commas between the items.

raidun99 replied:
I changed it, but there's still an issue somewhere. I'll keep staring at it to see if I can find it and fix it.

raidun99 replied:
I changed the exemplar, but I think there is still an issue. I'll keep staring at it.

sbillinge reviewed (commented):
please see the new inline comment.  The Travis stack-dump showed where, at least the first, problem lies.

sbillinge commented:
what error message are you getting from running `test_validate_python.py`?

On Thu, Jan 28, 2021 at 8:26 PM raidun99 <notifications@github.com> wrote:

> *@raidun99* commented on this pull request.
> ------------------------------
>
> In regolith/schemas.py
> <https://github.com/regro/regolith/pull/683#discussion_r566518941>:
>
> > @@ -1526,7 +1530,12 @@
>                           'initial project plan',
>              'audience': ['lead', 'pi', 'group_members'],
>              'status': 'proposed',
> -            'type': 'meeting'
> +            'type': 'meeting',
> +            'progress':[{
> +              - 'text': 'The samples have been synthesized and placed in the sample cupboard. '
>
> I changed the exemplar, but I think there is still an issue. I'll keep
> staring at it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#discussion_r566518941>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUNKDV2VEVRJLDZCPDLS4IFCVANCNFSM4WBLXWVA>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

raidun99 commented:
All I can see is
C:\Users\rubya\PycharmProjects\regolith\regolith\commands.py:234: SystemExit
and
short test summary info:
FAILED tests/test_validate.py::test_validate_python - SystemExit: Validation ...
============== 1 failed, 2 passed, 2 skipped, 1 warning in 6.95s ==============

sbillinge commented:
it looks as if you are running inside pycharm.  I don't have a lot of
experience there, but if you open a terminal, navigate to regolith
top-level directory and type `python run_tests.py` (possibly giving a `-vv`
tag, it should show a trace of what happened.

On Thu, Jan 28, 2021 at 9:59 PM raidun99 <notifications@github.com> wrote:

> All I can see is
> C:\Users\rubya\PycharmProjects\regolith\regolith\commands.py:234:
> SystemExit
> and
> short test summary info:
> FAILED tests/test_validate.py::test_validate_python - SystemExit:
> Validation ...
> ============== 1 failed, 2 passed, 2 skipped, 1 warning in 6.95s
> ==============
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-769541084>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUMUO436F5FW2LWZENTS4IQCTANCNFSM4WBLXWVA>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

raidun99 commented:
Okay, ran the new command from the terminal and got

tests/test_validate.py::test_validate_python
  C:\Users\rubya\PycharmProjects\regolith\regolith\schemas.py:4443: UserWarning: "" is not in the preferred entries for "recommendation", please consider changing this entry to conform or add this to the ``eallowed`` field in the schema.
    warn(
and
FAILED tests/test_validate.py::test_validate_python - SystemExit: Validation ...
FAILED tests/test_validators.py::test_exemplars[projecta] - assert False

It was marking  the progress field in the exemplar (something is off there), so I tried removing it from the exemplar and the schema. These same errors persisted - so multiple errors?

sbillinge commented:
did you run it with -vv.  I need to see the full stack dump.


On Thu, Jan 28, 2021 at 10:44 PM raidun99 <notifications@github.com> wrote:

> Okay, ran the new command from the terminal and got
>
> tests/test_validate.py::test_validate_python
> C:\Users\rubya\PycharmProjects\regolith\regolith\schemas.py:4443:
> UserWarning: "" is not in the preferred entries for "recommendation",
> please consider changing this entry to conform or add this to the eallowed
> field in the schema.
> warn(
> and
> FAILED tests/test_validate.py::test_validate_python - SystemExit:
> Validation ...
> FAILED tests/test_validators.py::test_exemplars[projecta] - assert False
>
> It was marking the progress field in the exemplar (something is off
> there), so I tried removing it from the exemplar and the schema. These same
> errors persisted - so multiple errors?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-769554653>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUPZGSMXJPRN532WYT3S4IVJXANCNFSM4WBLXWVA>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

raidun99 commented:
It's very long, I put it in a document. I hope that works okay

[FAILURES.docx](https://github.com/regro/regolith/files/5891531/FAILURES.docx)

sbillinge commented:
Hi Ruby,

It may be better to have a quick chat today and you can share your screen?
Better not to use MS word files around here, save as a text file, but more
importantly, that link doesn't work for me so I can't access the file.

S

On Thu, Jan 28, 2021 at 11:00 PM raidun99 <notifications@github.com> wrote:

> It's very long, I put it in a document. I hope that works okay
>
> FAILURES.docx
> <https://github.com/regro/regolith/files/5891531/FAILURES.docx>
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-769559055>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUKGYRWTKDMJVICAXFLS4IXFPANCNFSM4WBLXWVA>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

sbillinge commented:
Hi @raidun99 this is still failing CI.  It is a validation error.  YOu can find it yourself (good practice) but just for the record it seems to be this record in the exemplar:
```
Errors found in projecta

========================

ERROR in ly_newprojectum:

{'begin_date': ['must be of string type'], 'grants': ['must be of string type']}

datetime.date(2020, 4, 29)

['SymPy-1.1']

---------------
```
dates should be anyof type string date and grants, I think, should be a list, so the hte exemplar looks good in each case and it is the schema that should be changed I think.   Not sure why we missed this when we did the local testing.   

again, thank goodness for tests and CI....catching bugs.

raidun99 commented:
I've adjusted the schema and the exemplar (there were instances when the grant was a list and others when it was a string - I've changed it so that it is always a list)

Now I'm getting these failures that I'm unfamiliar with:
FAILED tests/test_builders.py::test_builder_python[html-fs] - assert ('<!DOCT...
FAILED tests/test_commands.py::test_fs_to_mongo -  FileNotFoundError: [WinError 2] The system cannot find the file specified

sbillinge commented:
is that the right fix?

On Sun, Jan 31, 2021 at 12:43 PM raidun99 <notifications@github.com> wrote:

> I've adjusted the schema and the exemplar (there were instances when the
> grant was a list and others when it was a string - I've changed it so that
> it is always a list)
>
> Now I'm getting these failures that I'm unfamiliar with:
> FAILED tests/test_builders.py::test_builder_python[html-fs] - assert
> ('<!DOCT...
> FAILED tests/test_commands.py::test_fs_to_mongo - FileNotFoundError:
> [WinError 2] The system cannot find the file specified
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-770420057>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUKSFQEEHA5BO6DHZCTS4WJERANCNFSM4WBLXWVA>
> .
>


-- 
Simon Billinge
Professor, Columbia University
Physicist, Brookhaven National Laboratory

raidun99 commented:
I'm not sure I know what you mean.

I could instead adjust the schema so that grants are anyof_type string, list. In the exemplar the grant is a string, and in tests/outputs files the grants are strings or lists depending on the example.

I reverted back to the commit from Friday but these two new failures are still present (as well as a validation error, but I think I know how to fix that one). Odd, I don't remember them on Friday.

sbillinge commented:
Here is the idea:
1) Q: what behavior do you want?
2) write a test that tests for that
3) write the code to pass the tests

So my question to you is to answer (1)


On Sun, Jan 31, 2021, 1:50 PM raidun99 <notifications@github.com> wrote:

> I'm not sure I know what you mean.
>
> I could instead adjust the schema so that grants are anyof_type string,
> list. In the exemplar the grant is a string, and in tests/outputs files the
> grants are strings or lists depending on the example.
>
> I reverted back to the commit from Friday but these two new failures are
> still present (as well as a validation error, but I think I know how to fix
> that one). Odd, I don't remember them on Friday.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-770430235>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUNGREUVGY54ZOZMA53S4WQ6PANCNFSM4WBLXWVA>
> .
>

raidun99 commented:
I think the grants field should be a list (which requires adjusting the tests slightly)
I'm thinking that a list is preferred over a string if a single projectum is funded by more than one grant. It's possible a single projectum could be funded by several grants, but looking through the group's projecta DB there is only one example of this.

However, every "grants" field in the projecta DB is a list, so it would be easier (for me) to change the schema so that grants is a list instead of changing the entire DB. Unless it is easier to access the information as a string, then in the long run it is easier to change the DB.

Thoughts?

sbillinge commented:
This is good analysis. A third option is to allow it to be anyof list or
string which wouldn't require changing the database or the schema.

Actually, we are the only people to use projecta so we don't have to worry
about backwards compatibility beyond our own databases.  I am not sure
whether your suggestion is better or allowing both. Allowing both makes the
code harder because you have to check type. Is that type checking code in
there now, where it is used?

On Sun, Jan 31, 2021, 6:22 PM raidun99 <notifications@github.com> wrote:

> I think the grants field should be a list (which requires adjusting the
> tests slightly)
> I'm thinking that a list is preferred over a string if a single projectum
> is funded by more than one grant. It's possible a single projectum could be
> funded by several grants, but looking through the group's projecta DB there
> is only one example of this.
>
> However, every "grants" field in the projecta DB is a list, so it would be
> easier (for me) to change the schema so that grants is a list instead of
> changing the entire DB. Unless it is easier to access the information as a
> string, then in the long run it is easier to change the DB.
>
> Thoughts?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-770470243>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUPE2TXQWFS5N7YDPI3S4XQ4JANCNFSM4WBLXWVA>
> .
>

raidun99 commented:
Yes, I've found in l_projectahelper.py the value of grants is checked if it is a list or a string, so either can be handled by the helper. (See line 167)

In a_projectum it makes grants a list, I think (See line 129). But as far as I can understand this is used to update the projecta, so as long as one of the accepted types is list it will continue to work.

So I think it is okay to allow both lists and strings, but my search was not exhaustive

sbillinge commented:
Let's go with that then

On Mon, Feb 1, 2021, 8:57 PM raidun99 <notifications@github.com> wrote:

> Yes, I've found in l_projectahelper.py the value of grants is checked if
> it is a list or a string, so either can be handled by the helper. (See line
> 167)
>
> In a_projectum it makes grants a list, I think (See line 129). But as far
> as I can understand this is used to update the projecta, so as long as one
> of the accepted types is list it will continue to work.
>
> So I think it is okay to allow both lists and strings, but my search was
> not exhaustive
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/683#issuecomment-771289772>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUJJVWE3KBVGMWK6CD3S45LYDANCNFSM4WBLXWVA>
> .
>

sbillinge commented:
haha, great job.....we got there!!!!!!
